### ISSUE TITLE
Speculation -spec-type=safety not working on one example. 

### ISSUE THREAD

**Issue by rasoolmaghareh** [tracer-x/klee#363]
In the master branch, speculation with -spec-type=safety and all the three strategies (timid, aggressive and custom) is not working on this example. However, when I run this program on the speculation branch speculation is performed successfully:

```
#include <klee/klee.h>

int main() {

    char b0,b1,b2;

    int x = 0;
    int y = 0,z=0, w=1;

    klee_make_symbolic(&b0, sizeof(char), "b0");
    klee_make_symbolic(&b1, sizeof(char), "b1");
    klee_make_symbolic(&b2, sizeof(char), "b2");

    if (b1 > 0) x = x + 3; else x = x + 0;
    if (b2 > 0) x = x + 6; else x = x + 0;
    if (b3 > 0) x = x + 9; else x = x + 0;
    int BOUND = 7;
    if (BOUND == x) {y++; z = z + 2; } else y--;
    if (x % 2 ==0 ) {w = w + 3; } else w = w + z;
    klee_assert( 2*y <= z && 3*y <= w);
    return 0;
}
```

For example, on the following attached c code, the speculation finishes quite fast:
```
KLEE: done: total instructions = 1880
KLEE: done: completed paths = 200, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     average branching depth of completed paths = 100
KLEE: done:     average branching depth of subsumed paths = 50.5
KLEE: done:     average instructions of completed paths = 931
KLEE: done:     average instructions of subsumed paths = 562.01
KLEE: done:     subsumed paths = 196
KLEE: done:     error paths = 0
KLEE: done:     program exit paths = 4
KLEE: done: generated tests = 3, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     error tests = 0
KLEE: done:     program exit tests = 3
```

While for the master branch after 30 seconds, I stopped the analysis with the following results:

```
KLEE: done: total instructions = 178201
KLEE: done: completed paths = 10814, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 90
KLEE: done:     average branching depth of completed paths = 96.0978
KLEE: done:     average branching depth of subsumed paths = 94.8565
KLEE: done:     average instructions of completed paths = 925.52
KLEE: done:     average instructions of subsumed paths = 879.476
KLEE: done:     subsumed paths = 8350
KLEE: done:     error paths = 0
KLEE: done:     program exit paths = 2374
KLEE: done: generated tests = 99, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 78
KLEE: done:     error tests = 0
KLEE: done:     program exit tests = 21
```
[spec.c.zip](https://github.com/tracer-x/klee/files/4424895/spec.c.zip)

**Comment by rasoolmaghareh** [comment]
Also @sanghu1790 has reported that the following program is slow on speculation with `-spec-strategy=custom -spec-type=safety`.

```
#include<stdio.h>
#ifdef LLBMC
#include <llbmc.h>
#else
#include <klee/klee.h>
#endif
typedef int bool;
int kappa  =    0;
char r1  ;
bool mode1  ;
bool mode2  ;
bool mode3  ;
void node1(void) 
{ 

  {
  if (mode1) {
    if ((int )r1 == 255) {
    kappa++;
      r1 = (char)2;
      
    }
    r1 = (char )((int )r1 + 1);
    mode1 = 0;
  }
  return;
}
}

void node2(void) 
{ 

  {
  if (mode2) {
    if ((int )r1 == 255) {
      r1 = (char)2;
    }
    r1 = (char )((int )r1 + 1);
    mode2 = 0;
    
  }
  return;
}
}

void node3(void) 
{ 
  {
  if (mode3) {
  
    if ((int )r1 == 255) {
      r1 = (char)2;
    }
    r1 = (char )((int )r1 + 1);
    mode3 = 0;
  }
  return;
}
}


int main(void) 
{ int i2 ;

  {
  klee_make_symbolic(& r1, sizeof(char ), "r1");
  klee_make_symbolic(& mode1, sizeof(bool ), "mode1");
  klee_make_symbolic(& mode2, sizeof(bool ), "mode2");
  klee_make_symbolic(& mode3, sizeof(bool ), "mode3");
  i2 = 0;
  while (i2 < 2) {
    node1();
    node2();
    node3();
    i2 = i2 + 1;
  }
  klee_assert(kappa < 1);
  //_SLICE(kappa);
  return (0);
}
}
```

**Comment by xuanlinhha** [comment]
@rasoolmaghareh on which branch of speculation do you run?
spec_aggressive
spec_custom
spec_custom_dyn
spec_timid

**Comment by rasoolmaghareh** [comment]
The Speculation branch:

[https://github.com/tracer-x/klee/tree/Speculation](url)

**Comment by xuanlinhha** [comment]
But we didn't work on that branch @rasoolmaghareh 
I remember we work and merge from branch beginning with `spec_`

**Comment by rasoolmaghareh** [comment]
@linh, I mean the old speculation branch. Can you please post the results of your comparison with master branch here?

**Comment by xuanlinhha** [comment]
[sp-bug3.c.txt](https://github.com/tracer-x/klee/files/4435908/sp-bug3.c.txt)
[info-master.txt](https://github.com/tracer-x/klee/files/4435910/info-master.txt)
[info-Speculation.txt](https://github.com/tracer-x/klee/files/4435913/info-Speculation.txt)

I see their running times are not much different on my machine

**Comment by sanghu1790** [comment]
Hi Rasool and Linh,

I experimented the program for the speculation slowness issue.
I ran "Mpals6-B2-cil.c" program which is big in size, but to see the difference
I found it suitable. Later we can reduce the size or take some other small program.

After my current investigation and recalling previous conversations, the issue of slowness
is due to running a program with and without providing "Spec_Avoid" files (these are dependent 
files after prepossessing).
I ran the programs with and without Spec_Avoid file on following commits:

1. master-0df7858a09741d0e19809dcc1a19fffd1179e445 
2. spec_custom_dyn-cfeb033c8912a8ce6ba84ed2b3f1117915fe40d3(most-recent)
3. spec_custom_dyn-ab4f9fb8cc0869e99a507b2512c97a505266b255(two-commits-back)

A. For the experiment with Spec_Avoid files on 1, 2, and 3, all finished in 
   35 sec approx. And also I see their stats are also same. But, some numbers
   in spec.txt files of all these three results are little changed, that I think 
   we can check, why it had happened.

B. For the experiment without Spec_Avoid files on 1 and 3 above are very slow (THE PROBLEM WE ARE DEALING),
   whereas for 2, it is very fast and finished in 34 sec approx, just like happened in A above. 
   Also, here the info stats and spec.txt files are very different. 

From the above investigation, I have observed that "Removing Independence checks" and without Spec_Avoid files
works very faster as compared to others. My question is why providing dependency files or not so much matter
in performance? If we provide dependency files then the checking time should be more so it should be more expensive,
but if we dont provide any such files so still it should be faster. At least the master branch should be align with 
spec_custom_dyn (most recent) branch. But, I remember this during the preparation of master branch we took the decision
to not to "Removing Independence checks" from speculation sub-tree. Now, we can discuss after looking into consequences. 

Still, @Linh, please check why our metrics stats of info and spec.txt files of these are different. Is there any mistake or
every thing is fine. 

@Rasool, Just to tell you, if you remember I encountored this issues because we were trying to run spec, with different number of 
Spec_Avoid files (like all files, no files, only one file, deleting one file) etc. And our observation was different performance on different runs.
Then we discussed and agreed that, if we dont provide any files still it should work fine, but this is not the case now in master branch.

All, files and results are in zip folder, have a look.
[Speculation-fix.zip](https://github.com/tracer-x/klee/files/4442574/Speculation-fix.zip)

I ran with these commands:

Run master branch
/home/sanghu/TracerX/tracerx/Release+Asserts/bin/klee --max-memory=32000 --max-time=600 -solver-backend=z3 --search=dfs -allow-external-sym-calls --watchdog -dump-states-on-halt=0 -write-BB-cov=4 -no-output -spec-dependency=/home/sanghu/Desktop/Speculation-fix -spec-strategy=custom -spec-type=coverage ${BENCHMARK}.bc

Run spec_custom_dyn branch
/home/sanghu/TracerX/tracerx/Release+Asserts/bin/klee --max-memory=32000 --max-time=600 -solver-backend=z3 --search=dfs -allow-external-sym-calls --watchdog -dump-states-on-halt=0 -write-BB-cov=4 -no-output -speculation ${BENCHMARK}.bc 

Thanks,
sanghu

**Comment by rasoolmaghareh** [comment]
> [sp-bug3.c.txt](https://github.com/tracer-x/klee/files/4435908/sp-bug3.c.txt)
> [info-master.txt](https://github.com/tracer-x/klee/files/4435910/info-master.txt)
> [info-Speculation.txt](https://github.com/tracer-x/klee/files/4435913/info-Speculation.txt)
> 
> I see their running times are not much different on my machine

I also checked and I was able to generate the same results. So this seems that there has not been any issue here. Let's now switch to the example problem that @sanghu has provided.

**Comment by rasoolmaghareh** [comment]
Summary of today's discussion:

The issue in the `-spec-type=safety` speculation is the dependence check and also the check for new basic blocks:

```  // stop when see the new BB
  llvm::BasicBlock *currentBB = current.txTreeNode->getBasicBlock();
  if (visitedBlocks.find(currentBB) == visitedBlocks.end()) {
    specFail++;
    speculativeBackJump(current, true);
    return StatePair(0, 0);
  }
```

These checks are not required for safety but are required for coverage.

The suggestion is as follows:

1- creating a new branch `speculation_safety_fix`
2- for the  `-spec-type=safety` option, we turn off these two checks.

**Comment by rasoolmaghareh** [comment]
@xuanlinhha: I have checked the `speculation_safety_fix` branch and it is showing correct results on our test program, please proceed to to generate a branch on master to apply these two changes for  `-spec-type=safety`

**Comment by rasoolmaghareh** [comment]
@xuanlinhha: I have checked the `speculation_safety_fix` branch and it is showing correct results on our test program, please proceed to to generate a branch on master to apply these two changes for  `-spec-type=safety`

**Comment by rasoolmaghareh** [comment]
Hi Linh,

I have checked the code. Speculation `-spec-type=safety` and `-spec-strategy=aggressive` is working fine. However, `-spec-strategy=timid` and `-spec-strategy=custom` are not working fine. 

The question I am trying to answer is that if we turn off the `dependence check` and also `new basic-block check`, then what is the difference between `aggressive` and, `timid` and `custom`? 

Can you tell me what is the difference between these three strategies when running with `-spec-type=safety`?

I am thinking about dropping strategy when doing  `-spec-type=safety`. 
May I know what you think @xuanlinhha @sanghu1790? 

Results for `-spec-type=safety` and aggressive:
```
KLEE: done: total instructions = 1096
KLEE: done: completed paths = 102, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     average branching depth of completed paths = 51
KLEE: done:     average branching depth of subsumed paths = 26
KLEE: done:     average instructions of completed paths = 588
KLEE: done:     average instructions of subsumed paths = 390.52
KLEE: done:     subsumed paths = 98
KLEE: done:     error paths = 0
KLEE: done:     program exit paths = 4
KLEE: done: generated tests = 3, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     error tests = 0
KLEE: done:     program exit tests = 3
```

Results for `-spec-type=safety` and timid:
```
KLEE: done: total instructions = 406495
KLEE: done: completed paths = 39016, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     average branching depth of completed paths = 49
KLEE: done:     average branching depth of subsumed paths = 37.5849
KLEE: done:     average instructions of completed paths = 604.946
KLEE: done:     average instructions of subsumed paths = 484.066
KLEE: done:     subsumed paths = 36570
KLEE: done:     error paths = 0
KLEE: done:     program exit paths = 2446
KLEE: done: generated tests = 51, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     error tests = 0
KLEE: done:     program exit tests = 51
```
Results for `-spec-type=safety` and custom:
```
KLEE: done: total instructions = 406495
KLEE: done: completed paths = 39016, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     average branching depth of completed paths = 49
KLEE: done:     average branching depth of subsumed paths = 37.5849
KLEE: done:     average instructions of completed paths = 604.946
KLEE: done:     average instructions of subsumed paths = 484.066
KLEE: done:     subsumed paths = 36570
KLEE: done:     error paths = 0
KLEE: done:     program exit paths = 2446
KLEE: done: generated tests = 51, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     error tests = 0
KLEE: done:     program exit tests = 51
```

**Comment by xuanlinhha** [comment]
If `-spec-type=safety`, I turned off 2 checkings for all strategies
```
if (SpecTypeToUse == SAFETY) {
        if (SpecStrategyToUse == TIMID) {
          StatsTracker::increaseEle(curBB, 1, true);
        } else if (SpecStrategyToUse == AGGRESSIVE) {
          // open speculation & result may be success or fail
          StatsTracker::increaseEle(curBB, 0, true);
          return addSpeculationNode(current, condition, binst, isInternal,
                                    true);
        } else if (SpecStrategyToUse == CUSTOM) {
          // open speculation & result may be success or fail and Now second
          // check
          if (specSnap[binst] != visitedBlocks.size()) {
            dynamicYes++;
            StatsTracker::increaseEle(curBB, 0, true);
            return addSpeculationNode(current, condition, binst, isInternal,
                                      true);
          } else {
            dynamicNo++;
            // then close speculation & do marking as deletion
            txTree->markPathCondition(current, unsatCore);
            return StatePair(&current, 0);
          }
        }
      } else {
        if (SpecStrategyToUse == TIMID) {
          std::set<std::string> vars = extractVarNames(current, binst);
          if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
            independenceYes++;
            StatsTracker::increaseEle(curBB, 0, true);
            StatsTracker::increaseEle(curBB, 2, false);
          } else {
            independenceNo++;
            StatsTracker::increaseEle(curBB, 1, true);
          }
        } else if (SpecStrategyToUse == AGGRESSIVE) {
          // check independency
          std::set<std::string> vars = extractVarNames(current, binst);
          if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
            // open speculation & assume success
            independenceYes++;
            StatsTracker::increaseEle(curBB, 0, true);
            StatsTracker::increaseEle(curBB, 2, false);
            return StatePair(&current, 0);
          } else {
            // open speculation & result may be success or fail
            independenceNo++;
            StatsTracker::increaseEle(curBB, 0, true);
            return addSpeculationNode(current, condition, binst, isInternal,
                                      true);
          }
        } else if (SpecStrategyToUse == CUSTOM) {
          // check independency
          std::set<std::string> vars = extractVarNames(current, binst);
          if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
            // open speculation & assume success
            independenceYes++;
            //          StatsTracker::increaseEle(curBB, 0, true);
            //          StatsTracker::increaseEle(curBB, 2, false);
            return StatePair(&current, 0);
          } else {
            // open speculation & result may be success or fail and Now second
            // check
            independenceNo++;
            if (specSnap[binst] != visitedBlocks.size()) {
              dynamicYes++;
              StatsTracker::increaseEle(curBB, 0, true);
              return addSpeculationNode(current, condition, binst, isInternal,
                                        true);
            } else {
              dynamicNo++;
              // then close speculation & do marking as deletion
              txTree->markPathCondition(current, unsatCore);
              return StatePair(&current, 0);
            }
          }
        }
      }
```

**Comment by sanghu1790** [comment]
@rasoolmaghareh I feel, type safety and coverage both are important for all versions timid, aggressive, and custom. Timid is more light weight and aggressive is more heavy weight. However, custom_dyn is reasonable because of choice of speculation. As you said, for safety "new basic-block check" is disabled for all three version which is correct. But, when we say "dependence check" turned off, it means we dont check the independence checks  of Spec_Avoid files(if supplied which I expect is required) within "speculationFork" however independence checks will be there in "branchFork". I suspect the the statistics you are getting different because of this incompatibility. 
@xuanlinhha May I request you to check in this fix branch for timid, aggressive, and custom, where,
1. We all time do independence check in branchFork and speculationFork? YES/NO
2. We do independence check in branchFork and not speculationFork for all three? YES/NO
3. We do independence check in branchFork and not speculationFork for timid and custom? YES/NO
4. We do independence check in branchFork and speculationFork for aggressive, and not for others?YES/NO

@rasoolmaghareh Rasool I feel before dropping the strategy and taking the decision lets get above answers from @xuanlinhha.
I remember, we took decision regarding the checks on/off of on branches, that's why statistics are different.

**Comment by rasoolmaghareh** [comment]
sure, let's wait for Linh's response.

**Comment by xuanlinhha** [comment]
Based on the last commit of this branch: `maste_speculation_safety_fix`
1. NO Because some cases we do not check independency in branchFork
2. NO Because we check independency for CUSTOM in speculationFork
3. NO Because in branchFork we do not check independency if type is safety
4. NO Because in branchFork, if type is safety, we also do not check independency for aggressive

**Comment by sanghu1790** [comment]
@xuanlinhha Thanks for answers. In conclusion, we dont do "independency check" in branchFork for all three timid, aggressive, and custom. We do "independency check" in speculationFork for CUSTOM only and not for timid and aggressive. I think something is wrong as per our original definitions we decided. 
My interpretation:
I always believed that for safety and coverage strategies we always do independence checks in branchFork, and I remember we had some choice to turn on/off the checks in speculationFork in all the three versions.  I am pretty sure that the difference we decided in between safety and coverage strategies was just only disabling the "new basic-block check" in safety thats all, and other features will be as it is same.  But, here the scenario is very different. What I can see is that for timid and aggressive there is no need to supply Spec_Avoid files entirely if safety is enabled, which we never decided to do. That's why all time I was saying that we need to supply Spec_Avoid files to all versions on both safety and coverage strategies. In performance, I remember safety would be little faster but may not give accurate block coverage, but for coverage strategy performance will be little slower and block coverage will be accurate (since if coverage enabled it does both safety and coverage).

@rasoolmaghareh Do you remember anything if we took such above decision? Can you see a lot of ambiguities? 
@rasoolmaghareh and @xuanlinhha I think we need to discuss on this.
Thanks,
Sanghu

**Comment by xuanlinhha** [comment]
This is in `branchFork`
```
if (INTERPOLATION_ENABLED && SpecTypeToUse != NO_SPEC &&
        TxSpeculationHelper::isStateSpeculable(current)) {
  ...
  if (SpecTypeToUse == SAFETY) {
    if (SpecStrategyToUse == TIMID) {
      StatsTracker::increaseEle(curBB, 1, true);
    } else if (SpecStrategyToUse == AGGRESSIVE) {
      // open speculation & result may be success or fail
      StatsTracker::increaseEle(curBB, 0, true);
      return addSpeculationNode(current, condition, binst, isInternal,
                                true);
    } else if (SpecStrategyToUse == CUSTOM) {
      // open speculation & result may be success or fail and Now second
      // check
      if (specSnap[binst] != visitedBlocks.size()) {
        dynamicYes++;
        StatsTracker::increaseEle(curBB, 0, true);
        return addSpeculationNode(current, condition, binst, isInternal,
                                  true);
      } else {
        dynamicNo++;
        // then close speculation & do marking as deletion
        txTree->markPathCondition(current, unsatCore);
        return StatePair(&current, 0);
      }
    }
  } else {
    if (SpecStrategyToUse == TIMID) {
      std::set<std::string> vars = extractVarNames(current, binst);
      if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
        independenceYes++;
        StatsTracker::increaseEle(curBB, 0, true);
        StatsTracker::increaseEle(curBB, 2, false);
      } else {
        independenceNo++;
        StatsTracker::increaseEle(curBB, 1, true);
      }
    } else if (SpecStrategyToUse == AGGRESSIVE) {
      // check independency
      std::set<std::string> vars = extractVarNames(current, binst);
      if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
        // open speculation & assume success
        independenceYes++;
        StatsTracker::increaseEle(curBB, 0, true);
        StatsTracker::increaseEle(curBB, 2, false);
        return StatePair(&current, 0);
      } else {
        // open speculation & result may be success or fail
        independenceNo++;
        StatsTracker::increaseEle(curBB, 0, true);
        return addSpeculationNode(current, condition, binst, isInternal,
                                  true);
      }
    } else if (SpecStrategyToUse == CUSTOM) {
      // check independency
      std::set<std::string> vars = extractVarNames(current, binst);
      if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
        // open speculation & assume success
        independenceYes++;
        //          StatsTracker::increaseEle(curBB, 0, true);
        //          StatsTracker::increaseEle(curBB, 2, false);
        return StatePair(&current, 0);
      } else {
        // open speculation & result may be success or fail and Now second
        // check
        independenceNo++;
        if (specSnap[binst] != visitedBlocks.size()) {
          dynamicYes++;
          StatsTracker::increaseEle(curBB, 0, true);
          return addSpeculationNode(current, condition, binst, isInternal,
                                    true);
        } else {
          dynamicNo++;
          // then close speculation & do marking as deletion
          txTree->markPathCondition(current, unsatCore);
          return StatePair(&current, 0);
        }
      }
    }
  }
}
```

This is in `speculationFork`
```
if (INTERPOLATION_ENABLED && SpecTypeToUse != NO_SPEC &&
    TxSpeculationHelper::isStateSpeculable(current)) {
  if (SpecStrategyToUse == TIMID) {
    // do nothing
  } else if (SpecStrategyToUse == AGGRESSIVE) {
    // open speculation & result may be success or fail
    return addSpeculationNode(current, condition, binst, isInternal, true);
  } else if (SpecStrategyToUse == CUSTOM) {

    std::set<std::string> vars = extractVarNames(current, binst);
    if (TxSpeculationHelper::isIndependent(vars, bbOrderToSpecAvoid)) {
      // open speculation & assume success
      //          independenceYes++;
      return StatePair(&current, 0);
    } else {
      //          independenceNo++;
      // open speculation & result may be success or fail
      if (specSnap[binst] != visitedBlocks.size()) {
        //            dynamicYes++;
        return addSpeculationNode(current, condition, binst, isInternal,
                                  true);
      } else {
        //            dynamicNo++;
        // then close speculation & do marking as deletion
        txTree->markPathCondition(current, unsatCore);
        return StatePair(&current, 0);
      }
    }
  }
}
```

**Comment by sanghu1790** [comment]
@xuanlinhha Thanks linh. I read the code and understood the coding part. And it is as same as what you mentioned and I concluded. At least we are on same page. I am fine with the coverage part which is my interpretation says, and I remember this. But for safety part, I am really not able to re-call our discussion. Still, I am thinking why we are doing in such a way. This is also, little unlikely in for safety strategy in branchFork for CUSTOM version, that you have "dynamic checks" without "independence checks". How it is useful, I am not able to understand?

@xuanlinhha Also, linh which part of code is for "new basic-block check" which is different in safety and coverage?

**Comment by xuanlinhha** [comment]
@sanghu1790 it is in `executeInstruction` function:
```
if (SpecTypeToUse == COVERAGE) {
      llvm::BasicBlock *currentBB = state.txTreeNode->getBasicBlock();
      if (visitedBlocks.find(currentBB) == visitedBlocks.end()) {
        if (specFailNew.find(pp) != specFailNew.end()) {
          specFailNew[pp] = specFailNew[pp] + 1;
        } else {
          specFailNew[pp] = 1;
        }
        // check interpolation at is program point
        bool hasInterpolation = TxSubsumptionTable::hasInterpolation(state);
        if (!hasInterpolation) {
          if (specFailNoInter.find(pp) != specFailNoInter.end()) {
            specFailNoInter[pp] = specFailNoInter[pp] + 1;
          } else {
            specFailNoInter[pp] = 1;
          }
        }
        // add to visited BB
        // This is disabled to not to count blocks in speculation subtree
        // visitedBlocks.insert(currentBB);
        specFail++;
        speculativeBackJump(state);
        return;
      }
    }
  }
```

**Comment by rasoolmaghareh** [comment]
This is our strategies for `spec-coverage`:
**Timid:** Check independence, if yes then open speculation & assume `success`. If dependent, then perform marking similar to deletion.
**Aggressive:** Check independence, if yes then open speculation & assume `success`. If dependent then open speculation and speculation's result may be `success` or `fail` (do marking if `fail`).
**Custom:** Check independence, if yes then open speculation & assume `success`. If no then do the second check (check if the coverage has changed since the last speculation at this node). If the second check is satisfied then open speculation & result may be `success` or `fail`; else if the second check is not satisfied then close speculation & do marking as deletion.

From these, it is obvious for me that the independence check should not be in safety mode since it is possible that an independence check might say yes for a subtree that contains memory errors (where speculation should have `failed`) and it reports `success`. Please let me know if you agree with this @xuanlinhha @sanghu1790 and I will then proceed to the second check.

**Comment by sanghu1790** [comment]
HI @rasoolmaghareh @xuanlinhha 
Thanks for your messages. 
@rasoolmaghareh from you last paragraph it is not clear that you are talking about branchFork or speculationFork or I assume entirely. Then there is an ambiguity. 
For easy understanding for us I have prepared a matrix in this file [Speculation - MATRIX.pdf](https://github.com/tracer-x/klee/files/4478778/Speculation.-.MATRIX.pdf)

If I consider your point @rasoolmaghareh for safety, then all the time we never need to supply any Spec_Avoid files. If we dont supply Spec_Avoid files then why in speculationFork we have "Independence Check" for CUSTOM.  BTW, @xuanlinhha you have not provided the option of "-spec-type" in speculationFork. The code is common for both safety and coverage. @rasoolmaghareh Regarding, "dynamic check" in branchFork and speculationFork I am fine, because I remember this decision we tool that without "Independence Check" we can do "dynamic check". @rasoolmaghareh @xuanlinhha If we are completely agree that in safety mode we dont required any Spec_Avoid file then we need to remove the checks from speculationFork when it is safety mode for CUSTOM. There is an ambiguity.  

Now for Coverage, I am pretty ok with the independence checks in branchFork for all three versions as coded. Now, currently we have only we have independence checks in CUSTOM not in TIMID and AGGRESSIVE, which is again ok as we decided. But, here I propose one thing that to implement one option of enabling/disabling independence check within speculationFork for TIMID, AGGRESSIVE, CUSTOM. Because I remember for some program disabling this check outperform and some time enabling this. So we can have both options to do.

**Comment by xuanlinhha** [comment]
so now do we need to change codes following the matrix from @sanghu1790 or you need to discuss with Joxan again?

**Comment by sanghu1790** [comment]
@xuanlinhha To make it more clear, the MATRIX I made from above the above code you posted. It is not my proposed one. We need to discuss on the definition and MATRIX.
 Whether the current code is aligned with the definitions or not

**Comment by xuanlinhha** [comment]
I see @sanghu1790

**Comment by rasoolmaghareh** [comment]
@xuanlinhha Can you please confirm that the Matrix is inline with the code?

**Comment by xuanlinhha** [comment]
yes, the matrix from @sanghu1790 is correct

**Comment by sanghu1790** [comment]
Thanks to both of you @rasoolmaghareh and @xuanlinhha . 
Now, It will be easy for us to fix the issue.

**Comment by rasoolmaghareh** [comment]
This is my suggestion @xuanlinhha @sanghu1790:

1- We drop independence check since it is unsound in `-spec-type=safety`. 
2- Since `Timid` in `-spec-type=safety` becomes very similar to `aggressive`, I suggest we drop `timid`:

![image](https://user-images.githubusercontent.com/9147962/79433522-cc591680-7fff-11ea-9256-35ee8e380a1a.png)

**Comment by sanghu1790** [comment]
@rasoolmaghareh Yes I do agree with you

@xuanlinhha Linh, lets fix this issues first by dropping timid in safety. 
Secondly. Kindly remove the "Independence check" from speculationFork when type is safety.
Third point is to provide one command option to enable/disable "Independence check" in speculationFork for coverage type as we discussed.

**Comment by rasoolmaghareh** [comment]
@xuanlinhha Let's proceed with safety first, then we will address the independence issue, under the issue for speculation coverage (issue #364)

**Comment by sanghu1790** [comment]
@rasoolmaghareh I agree.

**Comment by xuanlinhha** [comment]
Hi all,
I made a new commit according to the new matrix, can you please check again with the programs you have? Thanks

**Comment by rasoolmaghareh** [comment]
@xuanlinhha I ran now and it runs correctly for safety (on both aggressive and custom). I am merging the pull request (#365).

Fixed by #365.
